### PR TITLE
Add GrantType for JSON Web Token (JWT)

### PR DESCRIPTION
**Pull request by steffkes**
Hi Guys

For an API that is mainly used from a server w/o manual interaction i've implemented JWT-Handling. The Idea was to offer an API-Client which is based on guzzle - before discovering your plugin i thought of rolling my own, which is obviously not as nice as extending yours ;)

It's pretty basic (i.e. fitting [only] my needs) in some areas, so let me know what you think about it and where it probably needs to be extended / be a bit more flexible than it is right now.

-steffkes

### CODE REVIEW

**Comment by steffkes** [comment]
Ah, already forgot the relevant comment: i'm never completely sure where i'd place a dependency like `firebase/php-jwt` which is needed only for the `JwtBearer`.

it's completely not needed for everything, but _if_ you use JWT it's not only a suggestion it's a hard requirement. so i'm absolutely fine moving this dependency to a `suggest` entry or something like this - no hard feelings here!

**Comment by steffkes** [comment]
I didn't know that `SplFileObject::fread` requires at least `5.5.11` oO will see how we can work around that. since there is no specific version-requirement in `composer.json` i guess the travis-tests using `5.4` declare the minimum?

**Comment by pjcdawkins** [comment]
Yes, currently this plugin works with PHP 5.4 (5.3 is not compatible with Guzzle 4+), and it would be good to keep it that way. Could you just use `file_get_contents()`?

**Comment by pjcdawkins** [comment]
To answer your other question: `firebase/php-jwt` seems to be a very popular and actively maintained library - I think it's fine to include it as a dependency even if it isn't always going to be used.

**Comment by steffkes** [comment]
You can't read a `SplFileObject` with `file_get_contents` that doesn't work :) But i guess we could [use `eof` and `fgets` as described in the examples](http://php.net/manual/en/splfileobject.fgets.php#refsect1-splfileobject.fgets-examples) - i'll give that a try

**Comment by steffkes** [comment]
In general we could use `file_get_contents` to read a file, the idea was to ensure that we already get something that has to be a file (otherwise `SplFileObject`s Constructor barks already) and we don't have to do those checks on our own.

**Comment by steffkes** [comment]
Travis is happy now, Scrutinizer complains w/ the same message as before .. any idea @pjcdawkins about that? not sure if that can be safely ignored or even should ..?

**Comment by pjcdawkins** [comment]
@steffkes all green now :) thanks

Scrutinizer suggests some indentation/whitespace adjustments, and I agree: https://scrutinizer-ci.com/g/commerceguys/guzzle-oauth2-plugin/inspections/fc54ae2d-161f-4acd-bf2d-a52fe207b7b8/patches

The rest looks good, I'll review it fully ASAP. At the moment I think the tests should probably validate the JWT fully.

**Comment by steffkes** [comment]
Already on it :)

**Comment by pjcdawkins** [comment]
I expect the Scrutinizer error was/is due to this default timeout: https://scrutinizer-ci.com/docs/tools/external-code-coverage/#defining-timeout

**Comment by steffkes** [comment]
i see, thanks for letting me know. i've added an additional test to verify the `InvalidArgumentException` for the `private_key` gets thrown correctly, if needed.

**Comment by steffkes** [comment]
@pjcdawkins you let me know if this PR is ready to merge for you?

File: src/GrantType/JwtBearer.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    /**
+     * Compute JWT, signing with provided private key
+     */
+    protected function computeJwt()
+    {
+        $payload = [
+            'iss' => $this->config->get('client_id'),
+            'aud' => sprintf('%s/%s', rtrim($this->client->getBaseUrl(), '/'), ltrim($this->config->get('token_url'), '/')),
+            'exp' => time() + 60 * 60 * 1000,
```
Comment (pjcdawkins) [inline src/GrantType/JwtBearer.php]: 41 days - is this a normal/necessary expiry time?

**Comment by pjcdawkins** [comment]
@steffkes it's good, I have that question ^

I wonder if the `GrantTypeBase::getToken()` method should be refactored again, as the `getAdditionalOptions()` thing is a bit confusing. It may be simpler if I look at that after merging, though.

File: src/GrantType/JwtBearer.php (hunk 1 above)
Comment (steffkes) [inline src/GrantType/JwtBearer.php]: Um well, i guess that happens if you do too much JS - don't need to multiply by a thousand here .. an hour should be fine.

**Comment by steffkes** [comment]
> I wonder if the `GrantTypeBase::getToken()` method should be refactored again, as the `getAdditionalOptions()` thing is a bit confusing. It may be simpler if I look at that after merging, though.

it mainly depends on how you've setup your endpoints .. if we'd need to do (more or less) the same in case we would overwrite `getToken` in the specific class - if you have an idea, i'm up for the good stuff :)

the library is/was mainly based around `client_id`/`client_secret` - so the behavior of jwt doesn't fit right in .. that's probably obvious from the implementation ..

**Pull request merged**